cohortextractor/query_engines/base_sql.py

```diff
@@ -0,0 +1,687 @@
+
+    #
+    # DATABASE CONNECTION
+    #
+    @property
+    def engine(self):
+        if self._engine is None:
+            engine_url = sqlalchemy.engine.make_url(self.backend.database_url)
+            engine_url = set_driver(engine_url)
+            self._engine = sqlalchemy.create_engine(engine_url, future=True)
+        return self._engine
+
+    def create_output_group_tables(self):
+        """Queries to generate and populate interim tables for each output"""
+        # For each group of "output nodes" (roughly, "nodes which we know how
+        # to build a single select query for"), build a SQLAlchemy query to get
+        # their values.
+        for i, (group, output_nodes) in enumerate(self.output_groups.items()):
+            query = self.get_query_expression(group, output_nodes)
+            self.output_group_tables_queries[group] = query
+            # Create a Table object representing a temporary table into which
+            # we'll write the results of the query
+            table_name = self.get_temp_table_name(f"group_table_{i}")
+            columns = [
+                sqlalchemy.Column(c.name, c.type) for c in query.selected_columns
+            ]
+            table = sqlalchemy.Table(
+                table_name,
+                sqlalchemy.MetaData(),
+                *columns,
+            )
+            self.output_group_tables[group] = table
+
+    def create_codelist_tables(self):
+        """
+        For each codelist in use, build a SQLAlchemy representation of the
+        temporary table needed to store that codelist and then generate the
+        queries necessary to create and populate those tables
+        """
+        for n, codelist in enumerate(self.codelists):
+            codes = codelist.codes
+            max_code_len = max(map(len, codes))
+            collation = "Latin1_General_BIN"
+            table_name = self.get_temp_table_name(f"codelist_{n}")
+            table = sqlalchemy.Table(
+                table_name,
+                sqlalchemy.MetaData(),
+                sqlalchemy.Column(
+                    "code",
+                    sqlalchemy.types.String(max_code_len, collation=collation),
+                    nullable=False,
+                ),
+                sqlalchemy.Column(
+                    "system",
+                    sqlalchemy.types.String(6),
+                    nullable=False,
+                ),
+            )
+            self.codelist_tables[codelist] = table
+            # Constuct the queries needed to create and populate this table
+            self.codelist_tables_queries.append(sqlalchemy.schema.CreateTable(table))
+            # There's a limit of 999 on how many rows we can insert in one go using
+            # this method See:
+            # https://docs.microsoft.com/en-us/sql/t-sql/queries/table-value-constructor-transact-sql?view=sql-server-ver15#limitations-and-restrictions
+            for codes_batch in split_list_into_batches(codes, size=999):
```

Comment:
Did you consider moving this into the MSSQL engine? I can see that it'll work on other dbs, but it seems a bit odd here.

Reply:
Oh that's a good point. We should at least move the specific size limit to the MSSQL class. I'm also sure we're not done with the refactor here and there are still other MSSQL-isms hiding in the base class that we'll discover later.